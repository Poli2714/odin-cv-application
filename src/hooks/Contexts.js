import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
