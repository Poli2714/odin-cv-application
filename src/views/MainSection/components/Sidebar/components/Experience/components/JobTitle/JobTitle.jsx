import { useContext } from 'react';
import PropTypes from 'prop-types';

import { InputWrapper } from 'src/components/layout';
import { Label, TextInput } from 'src/components/forms';
import { ExperienceContext } from 'src/hooks/Contexts';

function JobTitle({ id }) {
  const { experiences, setExperiences } = useContext(ExperienceContext);

  const handleChangeJobTitle = (e, key) =>
    setExperiences({
      ...experiences,
      [key]: { ...experiences[key], jobTitle: e.target.value },
    });

  return (
    <InputWrapper>
      <Label labelName="Job Title" labelFor="jobTitle" />
      <TextInput
        autofocus={true}
        name="jobTitle"
        onChange={e => handleChangeJobTitle(e, id)}
        pattern="^\w+"
        placeholder="e.g. React developer"
        value={experiences[id].jobTitle}
      />
    </InputWrapper>
  );
}

JobTitle.propTypes = {
  id: PropTypes.string,
};

export default JobTitle;
