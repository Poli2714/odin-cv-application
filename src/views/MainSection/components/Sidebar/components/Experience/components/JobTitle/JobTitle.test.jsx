import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';
import PropTypes from 'prop-types';

import JobTitle from './JobTitle';
import { ExperienceContext } from 'src/hooks/Contexts';

vi.mock('src/components/layout', () => ({
  InputWrapper: vi.fn(({ children }) => (
    <div data-testid="input-wrapper">{children}</div>
  )),
}));

vi.mock('src/components/forms', () => ({
  Label: vi.fn(() => <label htmlFor="jobTitle">Mock label</label>),
  TextInput: vi.fn(({ onChange, value }) => (
    <input id="jobTitle" type="text" onChange={onChange} value={value} />
  )),
}));

function MockParentComponent({ id }) {
  const [experiences, setExperiences] = useState({
    testId: { jobTitle: '' },
    testId2: { jobTitle: 'Web Developer' },
  });

  return (
    <ExperienceContext.Provider value={{ experiences, setExperiences }}>
      <JobTitle id={id} />
    </ExperienceContext.Provider>
  );
}

MockParentComponent.propTypes = {
  id: PropTypes.string,
};

test('renders JobTitle', () => {
  render(<MockParentComponent id="testId" />);

  expect(screen.getByLabelText('Mock label')).toMatchSnapshot();
});

test('renders jobTitle with initial value', () => {
  render(<MockParentComponent id="testId2" />);

  expect(screen.getByLabelText('Mock label')).toHaveValue('Web Developer');
});

test('updates jobTitle value as a user types', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent id="testId" />);
  const input = screen.getByLabelText('Mock label');

  await user.type(input, 'Developer');
  expect(input).toHaveValue('Developer');
  await user.clear(input);
  expect(input).toHaveValue('');
  await user.type(input, 'Frontend Developer');
  expect(input).toHaveValue('Frontend Developer');
});
