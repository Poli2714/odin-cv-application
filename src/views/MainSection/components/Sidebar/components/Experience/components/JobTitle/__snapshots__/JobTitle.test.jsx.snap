// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders JobTitle 1`] = `
<input
  id="jobTitle"
  type="text"
  value=""
/>
`;
